packages/protocol/test-sol/devchain/e2e/common/EpochManager.t.sol

```diff
@@ -677,6 +679,205 @@ contract E2E_EpochManager_FinishNextEpochProcess is E2E_EpochManager {
+  function _calculateRevokeNeighbors(
+    address targetGroup,
+    uint256 revokeAmount
+  ) internal view returns (address lesser, address greater, uint index) {
+    (address[] memory groupsInOrder, GroupWithVotes[] memory groupWithVotesSimulated) = getGroupsWithVotes();
+
+    uint targetGroupIndexSimulated = groupWithVotesSimulated.length;
+    for (uint i = 0; i < groupWithVotesSimulated.length; i++) {
+      if (groupWithVotesSimulated[i].group == targetGroup) {
+        if (groupWithVotesSimulated[i].votes >= revokeAmount) {
+          groupWithVotesSimulated[i].votes -= revokeAmount;
+        } else {
+          groupWithVotesSimulated[i].votes = 0;
+        }
+        targetGroupIndexSimulated = i;
+        break;
+      }
+    }
+    require(targetGroupIndexSimulated < groupWithVotesSimulated.length, "Target group not found for revoke simulation");
+
+    sort(groupWithVotesSimulated);
+
+    lesser = address(0);
+    greater = address(0);
+    for (uint i = 0; i < groupWithVotesSimulated.length; i++) {
+      if (groupWithVotesSimulated[i].group == targetGroup) {
+        lesser = (i == groupWithVotesSimulated.length - 1) ? address(0) : groupWithVotesSimulated[i + 1].group;
+        greater = (i == 0) ? address(0) : groupWithVotesSimulated[i - 1].group;
+        break;
+      }
+    }
+
+    index = groupsInOrder.length;
+    for (uint i = 0; i < groupsInOrder.length; i++) {
+      if (groupsInOrder[i] == targetGroup) {
+        index = i;
+        break; 
+      }
+    }
+    require(index < groupsInOrder.length, "Target group not found for revoke index");
+  }
+
+  struct AliceContext {
+    address alice;
+    uint256 lockedAmount;
+    address targetGroup;
+  }
+
+  function _setupAlice() internal returns (AliceContext memory ctx) {
+    ctx.alice = vm.addr(uint256(keccak256(abi.encodePacked("alice"))));
+    vm.deal(ctx.alice, 100_000 ether);
+    ctx.lockedAmount = 1 ether;
+
+    vm.startPrank(ctx.alice);
+    accounts.createAccount();
+    lockedCelo.lock{ value: ctx.lockedAmount }();
+    vm.stopPrank();
+
+    uint256 actualLocked = lockedCelo.getAccountTotalLockedGold(ctx.alice);
+    require(actualLocked == ctx.lockedAmount, "Alice lock failed");
+
+    (, GroupWithVotes[] memory initialGroups) = getGroupsWithVotes();
+    require(initialGroups.length >= 3, "Not enough groups for test setup");
+    ctx.targetGroup = initialGroups[2].group;
+  }
+
+  function _aliceVote(AliceContext memory ctx) internal {
+    (address lesser, address greater) = _calculateVoteNeighbors(ctx.targetGroup, ctx.lockedAmount);
+    vm.prank(ctx.alice);
+    election.vote(ctx.targetGroup, ctx.lockedAmount, lesser, greater);
+    vm.stopPrank();
+  }
+
+  function _aliceActivate(AliceContext memory ctx) internal {
+    vm.startPrank(ctx.alice);
+    election.activate(ctx.targetGroup);
+    vm.stopPrank();
+    assertAlmostEqual(election.getActiveVotesForGroupByAccount(ctx.targetGroup, ctx.alice), ctx.lockedAmount, 10, "Alice activation failed");
```

Comment:
Why almost equal by `10`? 
What's the motivation for not using `assertEqual`?

Reply:
I added comment:

 // assert equal is used because usually locked celo != active votes since votes are adjusted by Celo inflation in function unitsToVotes

Originally I put there 10 so it has some wiggle room, but in reality 1 is enough. I could use assertEqual, but at least to me this seems a bit more readable since we would need to subtract 1 wei anyways

Reply:
Thanks for explanation 🙌